doc/TEMPLATE_ENGINE.md

```diff
@@ -0,0 +1,619 @@
+    const fragment = stringsToFragment.get(strings).cloneNode(true);
+    const targets = findTargets(fragment);
+    if (endNode) {
+      removeNodesBetween(node, endNode);
+      node.parentNode.insertBefore(fragment, endNode);
+    } else {
+      node.textContent = '';
+      node.appendChild(fragment);
+    }
+    nodeToContainerState.set(node, { targets, strings, endNode, items: [] });
+  }
+  const targets = nodeToContainerState.get(node).targets;
+  for (let iii = 0; iii < values.length; iii++) {
+    const target = targets[iii];
+    const value = values[iii];
+    const lastValue = target.value;
+    target.value = value;
+    switch (target.type) {
+      case 'attribute':
+        commitAttribute(target.node, target.name, value, lastValue);
+        break;
+      case 'boolean-attribute':
+        commitBooleanAttribute(target.node, target.name, value, lastValue);
+        break;
+      case 'property':
+        commitProperty(target.node, target.name, value, lastValue);
+        break;
+      case 'content':
+        commitContent(target.node, target.endNode, value, lastValue);
+        break;
+    }
+  }
+}
+
+// Tagged template function for idiomatic interpolation.
+function html(strings, ...values) {
+  const symbol = Symbol('__result__');
+  symbolToResult.set(symbol, { strings, values });
+  return symbol;
+}
+
+// Renderer to update DOM in the desired container.
+const render = (container, result) => {
+  commit(container, null, result);
+};
+
+export { html, render };
+```
+
+Phew! About ~220 lines now… but this is approaching something that could be
+valuable in real-world scenarios.
+
+**What it does**
+
+* It’s got the right functional interface (i.e., `html` and `render`).
+* Hydration is idiomatic JS.
+* It’s composable — i.e., you can nest, loop, etc.
+* Content can be interpolated between tags (e.g., text in a paragraph).
+* Attributes can be interpolated inside opening tags (e.g., flag on an item).
+* It understands html.
+* Boolean attributes and properties work via `.` / `?` DSL.
+* **DOM and HTML is cached (limited thrashing).** †
+* **All scripted implementation details are internalized.**
+* **Change-by-value detection.**
+
+**What it doesn’t do**
+
+* Some markup abstractions are leaking.
+* Many edge cases are not considered.
+
+† Note that the HTML caching actually hinges on a _very_ special property of
+  the `strings` argument in a tagged template function. When called from the
+  _exact same JS expression_ with the _exact same literals_, the value of the
+  `strings` argument will _always be exactly the same_. That means you can stick
+  it in a weak map.
+
+## Wrapping up!
+
+In just a couple iterations — we went from a naive implementation which just
+does recursive string concatenation… all the way to an html-aware, performant
+DOM management system. Nice!
+
+The remaining edge-cases and markup-cleanup won’t be covered here, but this
```

Comment:
perhaps "wont be covered here, but from here we would direct you to the source itself"